File: baboseira/javascript/abstract_factory.js
class Animal {
    constructor(nome, idade, nomeTutor, porte) {
        this.nome = nome;
        this.idade = idade;
        this.nomeTutor = nomeTutor;
        this.porte = porte;
    }

    comunicacao() {
        throw "Abstract Method";
    }
    comportamento() {
        throw "Abstract Method";
    }
}

class Cachorro extends Animal {
    comunicacao() {
        console.log(`Meu nome é ${this.nome} e meu/minha dono(a) é ${this.nomeTutor}.`);
    }
    comportamento() {
        console.log(`Gosto de latir e de brincar com meu/minha dono(a) ${this.nomeTutor}.`);
    }
}

class Gato extends Animal {
    comunicacao() {
        console.log(`Meu nome é ${this.nome} e meu/minha dono(a) é ${this.nomeTutor}.`);
    }
    comportamento() {
        console.log(`Gosto de dormir e de ignorar meu/minha dono(a) ${this.nomeTutor}.`);
    }
}

class Passaro extends Animal {
    comunicacao() {
        console.log(`Meu nome é ${this.nome} e meu/minha dono(a) é ${this.nomeTutor}.`);
    }
    comportamento() {
        console.log(`Gosto muito de cantar para meu/minha dono(a) ${this.nomeTutor}.`);
    }
}

class AnimalFactory {
    criarAnimal(nome, idade, nomeTutor, porte) {
        throw "Abstract Method";
    }
}

class FabricaCachorro extends AnimalFactory {
    criarAnimal(nome, idade, nomeTutor, porte) {
        return new Cachorro(nome, idade, nomeTutor, porte);
    }
}

class FabricaGato extends AnimalFactory {
    criarAnimal(nome, idade, nomeTutor, porte) {
        return new Gato(nome, idade, nomeTutor, porte);
    }
}

class FabricaPassaro extends AnimalFactory {
    criarAnimal(nome, idade, nomeTutor, porte) {
        return new Passaro(nome, idade, nomeTutor, porte);
    }
}

function criarExibirAnimal (factory, nome, idade, nomeTutor, porte) {
    const animal = factory.criarAnimal(nome, idade, nomeTutor, porte);
    animal.comunicacao();
    animal.comportamento();
}

criarExibirAnimal(new FabricaCachorro(), "Bebeto", 8, "Marcelo", "Médio");
criarExibirAnimal(new FabricaGato(), "Bolinha", 7, "Tifa", "Pequeno");
criarExibirAnimal(new FabricaPassaro(), "Juninho", 10, "Denise", "Médio");

File: sEngSoftware/abstractfactory.py
from abc import ABC, abstractmethod

class Animal(ABC):
    def __init__(self, nome, idade, nomeTutor, porte):
        self.nome = nome
        self.idade = idade
        self.nomeTutor = nomeTutor
        self.porte = porte
    @abstractmethod
    def comunicacao(self):
        pass
    @abstractmethod
    def comportamento(self):
        pass

class Cachorro(Animal):
    def comunicacao(self):
        print(f"Meu nome é {self.nome} e meu dono é {self.nomeTutor}.")
    def comportamento(self):
        print(f"Gosto de latir e de brincar com meu dono {self.nomeTutor}.")

class Gato(Animal):
    def comunicacao(self):
        print(f"Meu nome é {self.nome} e meu dono é {self.nomeTutor}.")
    def comportamento(self):
        print(f"Gosto de dormir e de ignorar meu dono {self.nomeTutor}.")

class Passaro(Animal):
    def comunicacao(self):
        print(f"Meu nome é {self.nome} e meu nome é {self.nomeTutor}.")
    def comportamento(self):
        print(f"Gosto muito de cantar para meu dono {self.nomeTutor}.")

class AnimalFactory(ABC):
    @abstractmethod
    def criarAnimal(self, nome, idade, nomeTutor, porte) -> Animal:
        pass

class FabricaCachorro(AnimalFactory):
    def criarAnimal(self, nome, idade, nomeTutor, porte):
        return Cachorro(nome, idade, nomeTutor, porte)

class FabricaGato(AnimalFactory):
    def criarAnimal(self, nome, idade, nomeTutor, porte):
        return Gato(nome, idade, nomeTutor, porte)

class FabricaPassaro(AnimalFactory):
    def criarAnimal(self, nome, idade, nomeTutor, porte):
        return Passaro(nome, idade, nomeTutor, porte)
    
def criarExibirAnimal(factory: AnimalFactory, nome, idade, nomeTutor, porte):
    animal = factory.criarAnimal(nome, idade, nomeTutor, porte)
    animal.comunicacao()
    animal.comportamento()

criarExibirAnimal(FabricaCachorro(), "Fred", 5, "Victor", "Médio")
criarExibirAnimal(FabricaGato(), "Cecília", 3, "Gabriela", "Pequeno")
criarExibirAnimal(FabricaPassaro(), "Zezé", 12, "Fabio", "Médio")